File: platform/python-sidecar/brahmagyan/bodha/bo_2-8.py
"""
brahmagyan/bodha/bo_2-8.py — holistic_bundle composite (L2 Bodha, BO-2-8)

B.11 whole-chart-read entrypoint. One call returns MSR signals + CGM graph edges +
CDLM domain links + RM resonance elements for a given chart_id — all 4 L2 Bodha
subsystems in a single, coherently provenance-wrapped response.

Architecture:
  - No new table. Pure aggregation over bodha_* tables populated by Wave-1.
  - DB queries run in parallel via asyncio; per-subsystem errors are isolated —
    one failure does not abort the bundle.
  - Returns the response schema:
      {
        signals: [...],
        graph_edges: [...],
        domain_links: [...],
        resonance: [...],
        provenance_envelope: {
          chart_id, queried_at, subsystems_ok, subsystems_errored,
          signal_count, graph_edge_count, domain_link_count, resonance_count,
          b11_floor_passed
        }
      }

BRAHMA-BO-2-8
"""

from __future__ import annotations

import asyncio
import logging
import os
from datetime import datetime, timezone
from typing import Any

import asyncpg

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# DB connection helper (reads DATABASE_URL / DIRECT_DATABASE_URL from env)
# ---------------------------------------------------------------------------

_DATABASE_URL: str | None = None


def _get_db_url() -> str:
    global _DATABASE_URL
    if _DATABASE_URL:
        return _DATABASE_URL
    for key in ("DATABASE_URL", "DIRECT_DATABASE_URL", "POSTGRES_URL"):
        val = os.environ.get(key)
        if val:
            _DATABASE_URL = val
            return val
    raise RuntimeError(
        "No database URL configured. Set DATABASE_URL, DIRECT_DATABASE_URL, or POSTGRES_URL."
    )


# ---------------------------------------------------------------------------
# Per-subsystem query functions
# ---------------------------------------------------------------------------


async def _fetch_signals(conn: asyncpg.Connection, chart_id: str) -> list[dict[str, Any]]:
    """Fetch MSR signals from bodha_signals for the given chart_id."""
    rows = await conn.fetch(
        """
        SELECT
            signal_id,
            domain,
            valence,
            confidence,
            salience,
            state,
            signal_text,
            source_citation,
            graha_refs,
            house_refs,
            created_at
        FROM bodha_signals
        WHERE chart_id = $1
        ORDER BY salience DESC NULLS LAST, signal_id
        """,
        chart_id,
    )
    return [dict(r) for r in rows]


async def _fetch_graph_edges(conn: asyncpg.Connection, chart_id: str) -> list[dict[str, Any]]:
    """Fetch CGM graph edges from bodha_graph_edges for the given chart_id."""
    rows = await conn.fetch(
        """
        SELECT
            edge_id,
            from_signal_id,
            to_signal_id,
            edge_type,
            weight,
            label,
            created_at
        FROM bodha_graph_edges
        WHERE chart_id = $1
        ORDER BY weight DESC NULLS LAST, edge_id
        """,
        chart_id,
    )
    return [dict(r) for r in rows]


async def _fetch_domain_links(conn: asyncpg.Connection, chart_id: str) -> list[dict[str, Any]]:
    """Fetch CDLM cross-domain linkages from bodha_domain_links for the given chart_id."""
    rows = await conn.fetch(
        """
        SELECT
            link_id,
            from_domain,
            to_domain,
            link_type,
            signal_ids,
            strength,
            note,
            created_at
        FROM bodha_domain_links
        WHERE chart_id = $1
        ORDER BY strength DESC NULLS LAST, link_id
        """,
        chart_id,
    )
    return [dict(r) for r in rows]


async def _fetch_resonance(conn: asyncpg.Connection, chart_id: str) -> list[dict[str, Any]]:
    """Fetch RM resonance elements from bodha_resonance for the given chart_id."""
    rows = await conn.fetch(
        """
        SELECT
            resonance_id,
            theme,
            signal_ids,
            resonance_type,
            amplitude,
            note,
            created_at
        FROM bodha_resonance
        WHERE chart_id = $1
        ORDER BY amplitude DESC NULLS LAST, resonance_id
        """,
        chart_id,
    )
    return [dict(r) for r in rows]


# ---------------------------------------------------------------------------
# Serialization helpers (convert asyncpg Row to plain JSON-serialisable dict)
# ---------------------------------------------------------------------------


def _serialise(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Convert any non-serialisable types (datetime, UUID, etc.) to strings."""
    out: list[dict[str, Any]] = []
    for row in rows:
        clean: dict[str, Any] = {}
        for k, v in row.items():
            if isinstance(v, datetime):
                clean[k] = v.isoformat()
            elif hasattr(v, "__str__") and not isinstance(v, (str, int, float, bool, type(None))):
                clean[k] = str(v)
            else:
                clean[k] = v
        out.append(clean)
    return out


# ---------------------------------------------------------------------------
# Main bundle function
# ---------------------------------------------------------------------------


async def holistic_bundle_async(chart_id: str) -> dict[str, Any]:
    """
    Async implementation. Parallel-fetches all 4 L2 Bodha subsystems and assembles
    the provenance-wrapped response envelope.

    Args:
        chart_id: UUID of the chart to read.

    Returns:
        {
            signals: [...],
            graph_edges: [...],
            domain_links: [...],
            resonance: [...],
            provenance_envelope: { ... }
        }
    """
    queried_at = datetime.now(timezone.utc).isoformat()
    subsystems_ok: list[str] = []
    subsystems_errored: list[str] = []

    signals: list[dict] = []
    graph_edges: list[dict] = []
    domain_links: list[dict] = []
    resonance: list[dict] = []

    db_url = _get_db_url()

    try:
        conn: asyncpg.Connection = await asyncpg.connect(db_url)
    except Exception as e:
        logger.error("[bo_2-8] DB connection failed: %s", e)
        return {
            "signals": [],
            "graph_edges": [],
            "domain_links": [],
            "resonance": [],
            "provenance_envelope": {
                "chart_id": chart_id,
                "queried_at": queried_at,
                "subsystems_ok": [],
                "subsystems_errored": ["MSR", "CGM", "CDLM", "RM"],
                "signal_count": 0,
                "graph_edge_count": 0,
                "domain_link_count": 0,
                "resonance_count": 0,
                "b11_floor_passed": False,
                "error": f"DB connection failed: {e}",
            },
        }

    async def safe(name: str, coro):
        try:
            result = await coro
            subsystems_ok.append(name)
            return result
        except Exception as e:
            logger.warning("[bo_2-8] subsystem %s failed: %s", name, e)
            subsystems_errored.append(name)
            return []

    try:
        results = await asyncio.gather(
            safe("MSR", _fetch_signals(conn, chart_id)),
            safe("CGM", _fetch_graph_edges(conn, chart_id)),
            safe("CDLM", _fetch_domain_links(conn, chart_id)),
            safe("RM", _fetch_resonance(conn, chart_id)),
        )
        signals, graph_edges, domain_links, resonance = results
    finally:
        await conn.close()

    signals = _serialise(signals)
    graph_edges = _serialise(graph_edges)
    domain_links = _serialise(domain_links)
    resonance = _serialise(resonance)

    # B.11 floor: at least MSR (signals) must be non-empty
    b11_floor_passed = len(signals) > 0

    return {
        "signals": signals,
        "graph_edges": graph_edges,
        "domain_links": domain_links,
        "resonance": resonance,
        "provenance_envelope": {
            "chart_id": chart_id,
            "queried_at": queried_at,
            "subsystems_ok": subsystems_ok,
            "subsystems_errored": subsystems_errored,
            "signal_count": len(signals),
            "graph_edge_count": len(graph_edges),
            "domain_link_count": len(domain_links),
            "resonance_count": len(resonance),
            "b11_floor_passed": b11_floor_passed,
        },
    }


def holistic_bundle(chart_id: str) -> dict[str, Any]:
    """
    Synchronous wrapper for use from FastAPI endpoints.

    Delegates to holistic_bundle_async via asyncio.run() when there is no running
    event loop, or via loop.run_until_complete() when there is.
    """
    try:
        loop = asyncio.get_event_loop()
        if loop.is_running():
            # Inside an async context (e.g. pytest-asyncio, FastAPI handler) —
            # create a new loop in a thread to avoid nesting.
            import concurrent.futures

            with concurrent.futures.ThreadPoolExecutor(max_workers=1) as pool:
                future = pool.submit(asyncio.run, holistic_bundle_async(chart_id))
                return future.result()
        else:
            return loop.run_until_complete(holistic_bundle_async(chart_id))
    except RuntimeError:
        return asyncio.run(holistic_bundle_async(chart_id))


# ---------------------------------------------------------------------------
# FastAPI router (mounted at /api/compute/brahma/bodha)
# ---------------------------------------------------------------------------


from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

router = APIRouter()


class HolisticBundleRequest(BaseModel):
    chart_id: str


class HolisticBundleResponse(BaseModel):
    ok: bool
    result: dict[str, Any]


@router.post("/holistic_bundle", response_model=HolisticBundleResponse)
async def holistic_bundle_endpoint(body: HolisticBundleRequest) -> HolisticBundleResponse:
    """
    POST /api/compute/brahma/bodha/holistic_bundle

    B.11 whole-chart-read entrypoint. Returns MSR signals + CGM graph edges +
    CDLM domain links + RM resonance elements for the given chart_id.

    Response: { ok: true, result: { signals, graph_edges, domain_links, resonance,
                                    provenance_envelope } }
    """
    if not body.chart_id or not body.chart_id.strip():
        raise HTTPException(status_code=400, detail="chart_id is required")

    bundle = await holistic_bundle_async(body.chart_id.strip())
    return HolisticBundleResponse(ok=True, result=bundle)

File: platform-mcp/src/tools/bo_2-8.ts
/**
 * bo_2-8.ts — holistic_bundle MCP tool (L2 Bodha, BRAHMA-BO-2-8)
 *
 * B.11 whole-chart-read entrypoint. One call returns all 4 L2 Bodha subsystems
 * (MSR signals + CGM graph edges + CDLM domain links + RM resonance) coherently.
 *
 * Tool name:  holistic_bundle
 * Input:      { chart_id: string }
 * Output:     { signals, graph_edges, domain_links, resonance, provenance_envelope }
 *
 * Architecture:
 *   - Calls the Python sidecar at POST /api/compute/brahma/bodha/holistic_bundle
 *   - Per-subsystem error isolation: one failed subsystem produces an empty array,
 *     not a bundle failure; provenance_envelope.subsystems_errored records failures.
 *   - No caching at this layer (caching deferred to BO-2-9 or platform layer).
 *   - Surgical: false — this IS the B.11 floor, not a bypass of it.
 *
 * BRAHMA-BO-2-8
 */

import { z } from 'zod'
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js'
import type { Principal } from '../types.js'
import { remoteAuthorize } from '../lib/authz.js'
import { budgetMcpContent } from '../lib/response_budget.js'

// ── Environment ────────────────────────────────────────────────────────────────

const PYTHON_SIDECAR_URL = (
  process.env['PYTHON_SIDECAR_URL'] ?? 'http://localhost:8001'
).replace(/\/$/, '')

const SIDECAR_API_KEY = process.env['PYTHON_SIDECAR_API_KEY'] ?? ''
const TOOL_TIMEOUT_MS = 30_000

// ── Response types ─────────────────────────────────────────────────────────────

export interface SignalRow {
  signal_id: string
  domain: string
  valence: string
  confidence: number | null
  salience: number | null
  state: string | null
  signal_text: string
  source_citation: string | null
  graha_refs: string[] | null
  house_refs: number[] | null
  created_at: string
}

export interface GraphEdgeRow {
  edge_id: string
  from_signal_id: string
  to_signal_id: string
  edge_type: string
  weight: number | null
  label: string | null
  created_at: string
}

export interface DomainLinkRow {
  link_id: string
  from_domain: string
  to_domain: string
  link_type: string
  signal_ids: string[] | null
  strength: number | null
  note: string | null
  created_at: string
}

export interface ResonanceRow {
  resonance_id: string
  theme: string
  signal_ids: string[] | null
  resonance_type: string
  amplitude: number | null
  note: string | null
  created_at: string
}

export interface ProvenanceEnvelope {
  chart_id: string
  queried_at: string
  subsystems_ok: string[]
  subsystems_errored: string[]
  signal_count: number
  graph_edge_count: number
  domain_link_count: number
  resonance_count: number
  b11_floor_passed: boolean
  error?: string
}

export interface HolisticBundleResult {
  signals: SignalRow[]
  graph_edges: GraphEdgeRow[]
  domain_links: DomainLinkRow[]
  resonance: ResonanceRow[]
  provenance_envelope: ProvenanceEnvelope
}

// ── Sidecar caller ─────────────────────────────────────────────────────────────

async function callSidecarHolisticBundle(chartId: string): Promise<HolisticBundleResult> {
  const url = `${PYTHON_SIDECAR_URL}/api/compute/brahma/bodha/holistic_bundle`

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(SIDECAR_API_KEY ? { 'X-API-Key': SIDECAR_API_KEY } : {}),
    },
    body: JSON.stringify({ chart_id: chartId }),
    signal: AbortSignal.timeout(TOOL_TIMEOUT_MS),
  })

  if (!response.ok) {
    const text = await response.text().catch(() => '(no body)')
    throw new Error(`Sidecar returned HTTP ${response.status}: ${text}`)
  }

  const json = (await response.json()) as { ok: boolean; result: HolisticBundleResult }
  if (!json.ok || !json.result) {
    throw new Error('Sidecar returned ok:false or missing result')
  }

  return json.result
}

// ── MCP tool registration ──────────────────────────────────────────────────────

/**
 * Register the holistic_bundle tool on the given McpServer instance.
 *
 * Called once per request in server.ts (stateless per-request server pattern).
 */
export function registerHolisticBundleTool(
  server: McpServer,
  principal: Principal
): void {
  server.tool(
    // Tool name
    'holistic_bundle',

    // Description (shown to the LLM)
    [
      'B.11 whole-chart-read entrypoint (Brahma L2 Bodha, BO-2-8).',
      'Returns all four L2 Bodha subsystems coherently in one call:',
      '  signals      — MSR (Multidimensional Signal Register): grounded signals per domain',
      '  graph_edges  — CGM (Chart Graph Model): valenced edges (reinforce/contradict/…)',
      '  domain_links — CDLM (Cross-Domain Linkage Matrix): cross-domain linkages',
      '  resonance    — RM (Resonance Map): thematic resonance elements',
      'provenance_envelope carries subsystem health + B.11 floor pass/fail flag.',
      'Use before any synthesis query that must read the whole chart.',
    ].join('\n'),

    // Input schema
    {
      chart_id: z
        .string()
        .min(1)
        .describe('UUID of the chart to read (e.g. the native chart UUID from the charts table)'),
    },

    // Handler
    async ({ chart_id }): Promise<{ content: Array<{ type: 'text'; text: string }> }> => {
      // M0 entitlement gate — require view access to the requested chart
      const authorized = await remoteAuthorize(principal, chart_id)
      if (!authorized) {
        return { content: [{ type: 'text', text: JSON.stringify({ error: 'AUTHZ_DENIED', chart_id }) }], isError: true } as never
      }

      const t0 = Date.now()
      let bundle: HolisticBundleResult

      try {
        bundle = await callSidecarHolisticBundle(chart_id)
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err)
        const errorBundle: HolisticBundleResult = {
          signals: [],
          graph_edges: [],
          domain_links: [],
          resonance: [],
          provenance_envelope: {
            chart_id,
            queried_at: new Date().toISOString(),
            subsystems_ok: [],
            subsystems_errored: ['MSR', 'CGM', 'CDLM', 'RM'],
            signal_count: 0,
            graph_edge_count: 0,
            domain_link_count: 0,
            resonance_count: 0,
            b11_floor_passed: false,
            error: message,
          },
        }
        const latency = Date.now() - t0
        const budgetedError = budgetMcpContent(
          {
            ok: false,
            tool: 'holistic_bundle',
            brahma_unit: 'BO-2-8',
            latency_ms: latency,
            result: errorBundle,
            error: message,
          },
          'holistic_bundle',
        )
        return {
          content: [
            {
              type: 'text',
              text: JSON.stringify(budgetedError, null, 2),
            },
          ],
        }
      }

      const latency = Date.now() - t0
      const envelope = bundle.provenance_envelope

      const budgeted = budgetMcpContent(
        {
          ok: true,
          tool: 'holistic_bundle',
          brahma_unit: 'BO-2-8',
          latency_ms: latency,
          b11_floor_passed: envelope.b11_floor_passed,
          summary: {
            signal_count: envelope.signal_count,
            graph_edge_count: envelope.graph_edge_count,
            domain_link_count: envelope.domain_link_count,
            resonance_count: envelope.resonance_count,
            subsystems_ok: envelope.subsystems_ok,
            subsystems_errored: envelope.subsystems_errored,
          },
          result: bundle,
        },
        'holistic_bundle',
      )

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(budgeted, null, 2),
          },
        ],
      }
    }
  )
}
